### PR TITLE
fix #2344 partial fix for #1753

### LINKED ISSUE #2344: Decaying item stops decaying when you move it
### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
      support matter (use https://otland.net/forums/support.16/ for support)
- [x] This issue is reproducible without changes to the code in this repository

### Steps to reproduce (include any configuration/script required to reproduce)
1. Make item decay, for example make new item in items.xml
2. item:decay()
3. move it to any equipment slot or any other position that it is in currently

This is the item I used:
```
<item id="28826" article="a" name="diamond arrow">
	<attribute key="weight" value="80" />
	<attribute key="slotType" value="ammo" />
	<attribute key="attack" value="37" />
	<attribute key="weaponType" value="ammunition" />
	<attribute key="ammoType" value="arrow" />
	<attribute key="shootType" value="diamondarrow" />
	<attribute key="maxHitChance" value="95" />
	<attribute key="decayTo" value="0" />
	<attribute key="duration" value="3600" />
	<attribute key="showduration" value="1" />
</item>
```

### Expected behaviour
Item should still be decaying

### Actual behaviour
Decaying is stopped.

### PR DESCRIPTION
**Pull request by EvilHero90**
- Items now get a decaystate check everytime they will be moved, added to another cylinder or transformed into another item.
-~~When leaving a house the items will still decay within the mailbox.~~ will still be started once the item is moved out of the mailbox.
- Decay will still stop after relog (depot/mailbox) but once the item gets moved out of either it starts normal again. This still needs a rework tho.

### CODE REVIEW

**Comment by pegason** [comment]
this is crashing the server via game_market (otclient) from the moment you purchase a groupable item and try to pull it from the inbox to stack.

**Comment by EvilHero90** [comment]
Do you have a gdb stacktrace by any chance?

**Comment by pegason** [comment]
> Do you have a gdb stacktrace by any chance?

Exception thrown: read access violation.
**this** was nullptr.

**Comment by EvilHero90** [comment]
@pegason can you try changing game.cpp from line 1214 to this:
```cpp
if (moveItem) {
	if (moveItem->getDuration() > 0) {
		if (moveItem->getDecaying() != DECAYING_TRUE) {
			moveItem->incrementReferenceCounter();
			moveItem->setDecaying(DECAYING_TRUE);
			g_game.toDecayItems.push_front(moveItem);
		}
	}
}
```
It's easier if you already know how to reproduce it, to see if it's fixed with this.

**Comment by Znote** [comment]
fix #2344 partial fix #1753

File: src/game.cpp [hunk 1]
```diff
@@ -1211,11 +1211,13 @@ ReturnValue Game::internalMoveItem(Cylinder* fromCylinder, Cylinder* toCylinder,
 		return retMaxCount;
 	}
 
-	if (moveItem->getDuration() > 0) {
-		if (moveItem->getDecaying() != DECAYING_TRUE) {
-			moveItem->incrementReferenceCounter();
-			moveItem->setDecaying(DECAYING_TRUE);
-			g_game.toDecayItems.push_front(moveItem);
+	if (moveItem) {
+		if (moveItem->getDuration() > 0) {
```
Comment (jo3bingham) [inline src/game.cpp]: I would have preferred `if (moveItem && moveItem->getDuration() > 0) {` to reduce nesting.

Comment (EvilHero90) [reply src/game.cpp]: sure, will do that

**Pull request merged**